gcpbuildcache/src/main/kotlin/androidx/build/gradle/gcpbuildcache/GcpStorageService.kt

```diff
@@ -174,16 +174,36 @@ internal class GcpStorageService(
             }
             return when (gcpCredentials) {
                 is ApplicationDefaultGcpCredentials -> {
-                    GoogleCredentials.getApplicationDefault().createScoped(scopes)
+                    val credentials = GoogleCredentials.getApplicationDefault().createScoped(scopes)
+                    try {
+                        credentials.refreshAccessToken()
```

Comment:
The reason this works is because the underlying `ServiceAccountCredentials` instance, supports `refresh`-ing a token even though its not expired yet. Typically clients should call: 

https://source.corp.google.com/piper///depot/google3/third_party/java_src/auth_kit/oauth2_http/java/com/google/auth/oauth2/OAuth2Credentials.java;rcl=545094018;l=175

If you wanted to make this credential type agnostic, then one thing you could consider is calling the `/tokenInfo` endpoint (https://developers.google.com/identity/sign-in/web/backend-auth#calling-the-tokeninfo-endpoint).
Using the `accessToken` you can make a request that looks something like:

```
https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=accessToken
```

This should return a `HTTP 200` if the token is valid. Otherwise, the payload will tell you exactly what went wrong. This endpoint also tell you all the claims in the credential, and perhaps that could be a part of the error message if anything else is unexpected.

An example Retrofit interface would look something like:

```kotlin
import com.google.gson.annotations.SerializedName
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.create
import retrofit2.http.GET
import retrofit2.http.Query

internal fun gson(config: GsonBuilder.() -> Unit = {}) {
  val builder = GsonBuilder()
  config.invoke(builder)
  builder.create()
}

data class TokenInfo(
    val audience: String,
    @SerializedName("user_id")
    val userId: String,
    // Space delimited list of scopes
    val scope: String,
    @SerializedName("expires_in")
    val expiresIn: Long
) {
    fun scopes(): List<String> {
        return scope.split(" ")
    }
}


interface TokenInfoService {
    @GET("/oauth2/v1/tokeninfo")
    fun tokenInfo(@Query("access_token") accessToken: String): Call<TokenInfo>

    companion object {
        fun tokenService(): TokenInfoService {
            val retrofit = Retrofit.Builder()
                .baseUrl("https://www.googleapis.com")
                .addConverterFactory(GsonConverterFactory.create(gson()))
                .build()

            return retrofit.create()
        }
    }
}
```

It's also okay to leave this as is with this caveat. You might want to add a comment explaining your reasoning if that's the case.

Reply:
I updated the refreshToken call to `refreshIfExpired` with clarifying comments.
When credentials have expired, accessToken won't be available to be sent to the tokenInfo endpoint. 
And refreshing the accessToken with expired credentials throws an exception, that we are attempting to catch and provide a more helpful error message.
Another issue with tokenInfo endpoint: `It is not suitable for use in production code as requests may be throttled or otherwise subject to intermittent errors`.
Since credentials expire for every 20h for a lot of users, I don't believe this would be a good choice here. So, just keeping the code as it for now.

Reply:
You can also validate the JWT yourself (without calling the `tokenInfo` endpoint). I guess that's what the documentation would like us to do. The Open ID configuration to use is:

https://accounts.google.com/.well-known/openid-configuration

You can extract the signature and validate it against the certs in the configuration. That is still a much better check than handling an exception from `refreshIfExpired`. Reason being, your credential may still not have expired (based on its `iat`, but the underlying refresh token would have been revoked). So `refreshIfExpired` won't handle that case for you.